osf/utils/caching.py

```diff
@@ -0,0 +1,63 @@
+"""
+A property cache mechanism.
+The cache is stored on the model as a protected attribute. Expensive
+property lookups, such as database access, can therefore be sped up
+when accessed multiple times in the same request.
+The property can also be safely set and deleted without interference.
+"""
+from __future__ import unicode_literals
+
+from functools import wraps
+
+# from https://github.com/etianen/django-optimizations/blob/master/src/optimizations/propertycache.py
```

Comment:
This code is BSD-3 licensed. Might need to put the license text in our `NOTICE` file: https://github.com/etianen/django-optimizations/blob/master/LICENSE